### ISSUE TITLE
Mapper Guided Temporary Read-Only Instances

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#909]
Many times when mapping read-only region requirements, the mapper wants to create an instance and have the instance only be made valid for the given task/copy/etc and then be immediately treated as invalid so that the instance can become eligible for garbage collection right away. I've heard various forms of this request from @ahenzinger, @manopapad, @magnatelee, and @alexaiken. 

This issue will track work on such a feature. This is not a hard feature to add. The main question is what the interface looks like in the mapper. How do people want to specify that an instance should be treated as "temporarily valid"?

Three caveats that I can already see with the implementation:
1. If two operations map the same region read-only with temporary validity, they will have no knowledge of each other and duplicate data movement will occur. I don't think there is anyway to avoid this easily and it will be up to the mapper to be aware that this can occur when using this feature.
2. This will only work with read-only privileges. I don't think there is anything analogous with read-write or reduce privileges.
3. The runtime might not be able to honor this in all cases. For example, if we're doing a read-only mapping after reductions, then it's like the runtime will need to hold onto the first instance where reductions are applied. Otherwise we can unsafe duplicate applications of reductions to the same instance if two different tasks/copies/ops map overlapping read-only regions to the same instance.

**Comment by streichler** [comment]
Isn't this a specific instance of the more general "when there are multiple valid copies of an instance, none of them are candidates for garbage collection" problem?  For example, when a region is produced in one memory and then a copy made in another memory for a read-only task, you'd like to be able to reclaim the one in the first memory (rather than having the one in the second memory be one-use) if you know subsequent usage will also prefer the second memory.

**Comment by lightsighter** [comment]
Yes, that is the more general problem. We could probably have a second mechanism that invalidates other valid instances if we want that. It would have many of the same caveats as the initial feature I mentioned at the top of the issue description.

**Comment by lightsighter** [comment]
Looking back at the code, I remembered we have a part of this feature for mapping tasks:

https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion/legion_mapping.h#L530

That allows you to specify read-only regions that you don't want to make "valid" after the task is done running. We're missing equivalence functionality on other operations like copies and dependent partitioning operations. We also don't have the feature that Sean asked for which is the ability to say to make the new copy the "only valid instance" after mapping and invalidate all prior ones. I think I would like to see some more code from people showing how they want to write such things in a mapper.

**Comment by lightsighter** [comment]
Work on new equivalence sets has delayed this to the next milestone.
